### DISCUSSION TITLE
has some problem in Type Inference

### DISCUSSION THREAD

**Discussion by candy-Tong** [vuejs/pinia#773] [Help and Questions] [answered]
### Reproduction

https://github.com/candy-Tong/pinia-demo

I set the `Ref<EdgeConfig>` type for the edge state. but in action, I use this.edge, the inferred type is not correct, it seems that it says that it reports an error
![image](https://user-images.githubusercontent.com/20377280/140311008-0161b780-d832-43ef-a142-e566d4697b2f.png)

### Steps to reproduce the behavior

1. yarn install
2. npm run tsc

![image](https://user-images.githubusercontent.com/20377280/140311301-62dce5e0-4e00-4a21-aa6d-3e401b8b991d.png)

### Expected behavior

there is no typescript error

### Actual behavior

has typescript error

**Comment by BenShelton** [comment, marked as answer]
A few things you can try:

1. The documentation for state says that type inference is more reliable when using an arrow function to define state, have you tried that? See https://pinia.esm.dev/core-concepts/state.html#state
2. Instead of using `shallowRef` you can simply just define the state similar to the following and it'll still be reactive:
```ts
state: () => ({
  node: undefined as NodeConfig | undefined,
  edge: undefined as EdgeConfig | undefined
}),
```
3. Within the action you can type cast the return value:
```ts
return this.edge as EdgeConfig | undefined
```

**Reply by posva** [reply]
I would personally go with just an arrow function for `state`. I tested and the type is correct.

**Reply by candy-Tong** [reply]
I test the arrow function,but it still not corret

**Comment by candy-Tong** [comment]
1. I test the arrow function,but it still not corret.( I had updated the code to [github](https://github.com/candy-Tong/pinia-demo))
![image](https://user-images.githubusercontent.com/20377280/140445098-07322670-1939-448c-9436-43ff95a5c118.png)

2. I want to use shallowRef because i don’t want it to be a deeply reactive object that is a third-party object.

3. Within the action you can type cast the return value. It's work for me. But that means I can't use `store.edge` directly and must use `store.getEdge` to get the right type
